src/VisualStudio/Core/Next/Remote/RemotableDataJsonRpcEx.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Remote;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    /// <summary>
+    /// Communication channel between remote host client and remote host.
+    /// 
+    /// this is framework's back channel to talk to remote host
+    /// 
+    /// for example, this will be used to deliver missing remotable data to remote host.
+    /// 
+    /// all connection will share one remotable data channel
+    /// </summary>
+    internal class RemotableDataJsonRpcEx : JsonRpcEx
+    {
+        private readonly IRemotableDataService _remotableDataService;
+        private readonly CancellationTokenSource _source;
+
+        /// <summary>
+        /// number of shared count
+        /// </summary>
+        private readonly object _gate;
+        private int _count;
```

Comment:
@sharwell hmm.. I dont actually agree. it ReferenceCountedDisposalbe require people who use it to know exactly how it is supposed to be used.